core/src/hash_table/mod.rs

```diff
@@ -34,54 +38,170 @@ pub trait HashTable: Send + Sync + Clone + 'static {
-        self.assert_pair_meta(&PairMeta::new(keys, &old_pair, LINK_NAME, &new_pair.key()))
+        self.assert_meta(&EntryMeta::new(
+            &keys.node_id(),
+            &old.key(),
+            LINK_NAME,
+            &new.key(),
+        ))
     }
 
-    /// set the status of a Pair to DELETED
-    fn retract_pair(&mut self, keys: &Keys, pair: &Pair) -> Result<(), HolochainError> {
-        self.assert_pair_meta(&PairMeta::new(
-            keys,
-            &pair,
+    /// Remove an Entry from the HashTable by flagging it DELETED
+    fn retract_entry(&mut self, keys: &Keys, entry: &Entry) -> Result<(), HolochainError> {
+        // Set the crud-status EntryMeta to DELETED
+        self.assert_meta(&EntryMeta::new(
+            &keys.node_id(),
+            &entry.key(),
             STATUS_NAME,
             &CrudStatus::DELETED.bits().to_string(),
         ))
     }
 
-    // meta
-    /// assert a given PairMeta in the HashTable
-    fn assert_pair_meta(&mut self, meta: &PairMeta) -> Result<(), HolochainError>;
+    /// Add link metadata to an Entry
+    fn add_link(&mut self, link: &Link) -> Result<(), HolochainError> {
+        // Retrieve entry from HashTable
+        let base_entry = self.entry(&link.base())?;
+        if base_entry.is_none() {
+            return Err(HolochainError::ErrorGeneric(
+                "Entry from base not found".to_string(),
+            ));
+        }
+        let base_entry = base_entry.unwrap();
+
+        // pre-condition: linking must only work on AppEntries
+        if base_entry.is_sys() {
+            return Err(HolochainError::InvalidOperationOnSysEntry);
+        }
+
+        // Retrieve LinkListEntry
+        let maybe_meta = self.meta_from_request(base_entry.key(), &link.to_attribute_name())?;
+        // Update or Create LinkListEntry
+        let new_meta: EntryMeta;
+        match maybe_meta {
+            // None found so create one
+            None => {
+                // Create new LinkListEntry & Entry
+                let lle = LinkListEntry::new(&[link.clone()]);
+                let new_entry = lle.to_entry();
+                // Add it to HashTable
+                self.put_entry(&new_entry)?;
+
+                // TODO #281 - should not have to create Keys
+                let key_fixme = ::agent::keys::Key::new();
+                let keys_fixme = Keys::new(&key_fixme, &key_fixme, "FIXME");
 
-    /// lookup a PairMeta from the HashTable by PairMeta key
-    fn pair_meta(&mut self, key: &str) -> Result<Option<PairMeta>, HolochainError>;
-    /// lookup all PairMeta for a given Pair
-    fn metas_for_pair(&mut self, pair: &Pair) -> Result<Vec<PairMeta>, HolochainError>;
+                // Create Meta
+                new_meta = EntryMeta::new(
+                    &keys_fixme.node_id(),
+                    &base_entry.key(),
+                    &link.to_attribute_name(),
+                    &new_entry.key(),
```

Comment:
So, first I was wondering why you are creating keys here. I see that EntryMeta needs them as source, which makes sense. But it does not make sense at all to just create some keys here for that.

When I glanced over this code some weeks back and approved, I was still under the impression that "hash_table" is meant to become the DHT module.

Now, taking into account that this hash_table is a building block for building chains and DHT shards.. (a CAS), I am very skeptical about putting this code in here.. and the fact that you have to conjure up a key for the source of this entry I take as confirmation for that doubt.

Which makes want to check the reducer for all of this. But there is none?!